python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,286 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Dict, Generic, Iterable, List, TypeVar
+
+from iceberg.types import (
+    IcebergType,
+    ListType,
+    MapType,
+    NestedField,
+    PrimitiveType,
+    StructType,
+)
+
+T = TypeVar("T")
+
+
+class Schema:
+    """A table Schema"""
+
+    def __init__(self, *columns: Iterable[NestedField], schema_id: int, identifier_field_ids: List[int] = []):
+        self._struct = StructType(*columns)  # type: ignore
+        self._schema_id = schema_id
+        self._identifier_field_ids = identifier_field_ids
+        self._name_index: Dict[str, int] = index_by_name(self)
+        self._id_index: Dict[int, NestedField] = {}  # Will be lazily set when self.id_index property method is called
+
+    def __str__(self):
+        return "table { \n" + "\n".join([" " + str(field) for field in self.columns]) + "\n }"
+
+    def __repr__(self):
+        return f"Schema(fields={repr(self.columns)})"
+
+    @property
+    def columns(self) -> Iterable[NestedField]:
+        return self._struct.fields
+
+    @property
+    def id(self) -> int:
+        return self._schema_id
+
+    @property
+    def identifier_field_ids(self) -> List[int]:
+        return self._identifier_field_ids
+
+    @property
+    def id_index(self) -> Dict[int, NestedField]:
+        if not self._id_index:
+            self._id_index = index_by_id(self)
+        return self._id_index
+
+    @property
+    def name_index(self) -> Dict[str, int]:
+        return self._name_index
+
+    def as_struct(self) -> StructType:
+        return self._struct
+
+    def find_field(self, name_or_id: str | int, case_sensitive: bool = True) -> NestedField:
+        if isinstance(name_or_id, int):
+            return self.id_index[name_or_id]
+        if case_sensitive:
+            field_id = self.name_index[name_or_id]
+        else:
+            name_index_lower = {name.lower(): field_id for name, field_id in self.name_index.items()}
```

Comment:
I don't think this class should ever iterate through a dictionary, except for just one time to produce the lower case name to id index. I think this should create the index and store it as an instance variable.